### PR TITLE
cmd/tailscale/cli: add login and switch subcommands

### PR DESCRIPTION
**Pull request by maisem**
```
USAGE
  login [flags]

"tailscale login" logs in this machine to your Tailscale network.

FLAGS
  --accept-dns, --accept-dns=false
    	accept DNS configuration from the admin panel (default true)
  --accept-routes, --accept-routes=false
    	accept routes advertised by other Tailscale nodes (default false)
  --advertise-exit-node, --advertise-exit-node=false
    	offer to be an exit node for internet traffic for the tailnet (default false)
  --advertise-routes string
    	routes to advertise to other nodes (comma-separated, e.g. "10.0.0.0/8,192.168.0.0/24") or empty string to not advertise routes
  --advertise-tags string
    	comma-separated ACL tags to request; each must start with "tag:" (e.g. "tag:eng,tag:montreal,tag:ssh")
  --auth-key string
    	node authorization key; if it begins with "file:", then it's a path to a file containing the authkey
  --exit-node string
    	Tailscale exit node (IP or base name) for internet traffic, or empty string to not use an exit node
  --exit-node-allow-lan-access, --exit-node-allow-lan-access=false
    	Allow direct access to the local network when routing traffic via an exit node (default false)
  --hostname string
    	hostname to use instead of the one provided by the OS
  --login-server string
    	base URL of control server (default https://controlplane.tailscale.com)
  --operator string
    	Unix username to allow to operate on tailscaled without sudo
  --qr, --qr=false
    	show QR code for login URLs (default false)
  --shields-up, --shields-up=false
    	don't allow incoming connections (default false)
  --ssh, --ssh=false
    	run an SSH server, permitting access per tailnet admin's declared policy (default false)
  --timeout duration
    	maximum amount of time to wait for tailscaled to enter a Running state; default (0s) blocks forever (default 0s)
```

```
USAGE
  tailscale switch <name>
  tailscale switch --list

"tailscale switch" switches between logged in profiles
```
Updates #713

cc @mihaip @willnorris @mayakacz

### CODE REVIEW

**Comment by mayakacz** [comment]
* To switch, would you do `tailscale profile --switch home` or just `tailscale profile home`? I guess I'm not sure how `switch` would work.
* You also proposed `rename` in the doc. Did you forgo that here in favour of `delete`?
* How does `create` work with current state? Do I run `tailscale profile --create` and it creates a _blank_ profile? do I need to pass it an email? does it create a profile based on my current logged in user?

**Comment by mayakacz** [comment]
I believe this is not in line with what's now in https://docs.google.com/document/d/1HfUIN-teHjniUzntwwtNQ_GZvu72YbN1aWAXxgPvlzk/edit#

**Comment by bradfitz** [comment]
> `--list, --list=false`

Hah, "`--list=false`" wrote nobody ever.

**Comment by bradfitz** [comment]
So we're going to have three commands (up, set, login) with all the prefs flags, eh?

File: cmd/tailscale/cli/switch.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+	"github.com/peterbourgon/ff/v3/ffcli"
+	"tailscale.com/ipn"
+)
+
+var switchCmd = &ffcli.Command{
+	Name:       "switch",
+	ShortHelp:  "Switches to a different Tailscale profile",
+	LongHelp:   `"tailscale switch" switches between logged in profiles`,
+	ShortUsage: "switch [<name>] [--list]",
```
Comment (bradfitz) [inline cmd/tailscale/cli/switch.go:22]: It's more like `"switch --list\nswitch <name>\n"` (you can do multiple lines, right?)

File: cmd/tailscale/cli/switch.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+	}
+	if profID == cp.ID {
+		printf("Already on profile %q\n", args[0])
+		os.Exit(0)
+	}
+	if err := localClient.SwitchProfile(ctx, profID); err != nil {
+		printf("Failed to switch to profile: %v\n", err)
+		os.Exit(1)
+	}
+	printf("Switched to profile %q\n", args[0])
```
Comment (bradfitz) [inline cmd/tailscale/cli/switch.go:82]: Switched or Switching? Looks like it might still fail based on following for loop.

**Comment by maisem** [comment]
> So we're going to have three commands (up, set, login) with all the prefs flags, eh?

I was hoping we can start hiding the pref flags from `up` in the next release and start steering users towards `tailscale login` to login and `tailscale set` to toggle prefs. We would leave `tailscale up` as just the analog to `tailscale down`

Another option would be to make `tailscale login` only have `--qr`, `--control-url`, `--authkey` and `--advertise-tags`.

**Comment by bradfitz** [comment]
> Another option would be to make tailscale login only have --qr, --control-url, --authkey and --advertise-tags.

Maybe we add `--up=BOOL` (up, start, run?) defaulting to false, so people can say `--up=false` to login, and then change settings with `tailscale set`, and then `tailscale up`, if they want to avoid a time window where they would've been up an running with the wrong settings (like, say, shields being down)

**Review by mihaip — APPROVED** [review]

File: client/tailscale/localclient.go [hunk 3]
```diff
@@ -903,3 +903,39 @@ func (r jsonReader) Read(p []byte) (n int, err error) {
...
+	}
+	body, err = lc.send(ctx, "GET", "/localapi/v0/profiles/", 200, nil)
+	if err != nil {
+		return
+	}
+	all, err = decodeJSON[[]ipn.LoginProfile](body)
+	return current, all, err
+}
+
+// NewProfile creates and switches to a new profile.
```
Comment (bradfitz) [inline client/tailscale/localclient.go:925]: say more that it's unnamed, not persisted, etc? when does it get saved? etc

Comment (bradfitz) [reply client/tailscale/localclient.go:925]: like say why this doesn't return a ProfileID

File: client/tailscale/localclient.go [hunk 4]
```diff
@@ -903,3 +903,39 @@ func (r jsonReader) Read(p []byte) (n int, err error) {
...
+	return err
+}
+
+// SwitchProfile switches to the given profile.
+func (lc *LocalClient) SwitchProfile(ctx context.Context, profile ipn.ProfileID) error {
+	_, err := lc.send(ctx, "POST", "/localapi/v0/profiles/"+url.PathEscape(string(profile)), 204, nil)
+	return err
+}
+
+// DeleteProfile removes the profile with the given name.
```
Comment (bradfitz) [inline client/tailscale/localclient.go:937]: given ID, no?

File: cmd/tailscale/cli/up.go [hunk 5]
```diff
@@ -117,7 +120,14 @@ func newUpFlagSet(goos string, upArgs *upArgsT) *flag.FlagSet {
 		upf.BoolVar(&upArgs.forceDaemon, "unattended", false, "run in \"Unattended Mode\" where Tailscale keeps running even after the current GUI user logs out (Windows-only)")
 	}
 	upf.DurationVar(&upArgs.timeout, "timeout", 0, "maximum amount of time to wait for tailscaled to enter a Running state; default (0s) blocks forever")
-	registerAcceptRiskFlag(upf, &upArgs.acceptedRisks)
+
+	if cmd == "up" {
```
Comment (bradfitz) [inline cmd/tailscale/cli/up.go:124]: not doing the `--up=false` thing for "login", eh? (even with all the prefs flags too)

Comment (maisem) [reply cmd/tailscale/cli/up.go:124]: I was going to get back to this as it requires some a bunch more testing on when the login happens and when we run stop.

**Comment by mayakacz** [comment]
> --auth-key string

Please confirm that `--authkey` also works.

Any reason to exclude (or maybe just not noted): `--netfilter-mode`, `--snat-subnet-routes` and `--unattended`?

How do you set name? Is it only with `tailscale set --name` (which I didn't think we were doing yet)? I don't see `tailscale login --name` listed.

**Comment by maisem** [comment]
> > --auth-key string
> 
> Please confirm that `--authkey` also works.

It does


> 
> Any reason to exclude (or maybe just not noted): `--netfilter-mode`, `--snat-subnet-routes` and `--unattended`?

Those are there, just not visible cause I was running this on macOS where we do not have those options.

> 
> How do you set name? Is it only with `tailscale set --name` (which I didn't think we were doing yet)? I don't see `tailscale login --name` listed.

This is going to be a follow up change.

File: client/tailscale/localclient.go (hunk 4 above)
Comment (maisem) [reply client/tailscale/localclient.go:937]: yep.

**Review by mayakacz — APPROVED** [review]

File: client/tailscale/localclient.go [hunk 6]
```diff
@@ -903,3 +903,43 @@ func (r jsonReader) Read(p []byte) (n int, err error) {
...
+		return
+	}
+	all, err = decodeJSON[[]ipn.LoginProfile](body)
+	return current, all, err
+}
+
+// NewProfile creates and switches to a new unnamed profile. The new profile is
+// not assigned an ID until it is persisted after a successful login.
+// In order to login to the new profile, the user must call LoginInteractive.
+func (lc *LocalClient) NewProfile(ctx context.Context) error {
```
Comment (bradfitz) [inline client/tailscale/localclient.go:928]: NewT makes it sounds like it returns T.

How about `SwitchToEmptyProfile` or something?

File: client/tailscale/localclient.go [hunk 7]
```diff
@@ -903,3 +903,43 @@ func (r jsonReader) Read(p []byte) (n int, err error) {
...
+
+// SwitchProfile switches to the given profile.
+func (lc *LocalClient) SwitchProfile(ctx context.Context, profile ipn.ProfileID) error {
+	_, err := lc.send(ctx, "POST", "/localapi/v0/profiles/"+url.PathEscape(string(profile)), 204, nil)
+	return err
+}
+
+// DeleteProfile removes the profile with the given ID.
+// If the profile is the current profile, an empty profile
+// will be selected as if NewProfile was called.
```
Comment (bradfitz) [inline client/tailscale/localclient.go:941]: as if SwitchToEmptyProfile (or whatever)

File: cmd/tailscale/cli/cli.go [hunk 8]
```diff
@@ -183,8 +183,12 @@ change in the future.
 		}
 	}
 	if envknob.UseWIPCode() {
-		rootCmd.Subcommands = append(rootCmd.Subcommands, idTokenCmd)
-		rootCmd.Subcommands = append(rootCmd.Subcommands, serveCmd)
+		rootCmd.Subcommands = append(rootCmd.Subcommands,
+			idTokenCmd,
+			serveCmd,
+			loginCmd,
```
Comment (bradfitz) [inline cmd/tailscale/cli/cli.go:189]: Let's make these like "debug": present by default if you know they're there.

And then if the plan is for them to be alpha, then document them in their ShortHelp + LongHelp loudly that they're alpha. That's what @shayne is doing in his ingress CLI PR.

File: cmd/tailscale/cli/login.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/peterbourgon/ff/v3/ffcli"
+)
+
+var loginArgs upArgsT
+
+var loginCmd = &ffcli.Command{
+	Name:       "login",
+	ShortUsage: "login [flags]",
+	ShortHelp:  "Log in to a Tailscale account",
+	LongHelp:   `"tailscale login" logs in this machine to your Tailscale network.`,
```
Comment (bradfitz) [inline cmd/tailscale/cli/login.go:20]: "logs this machine in to " ?

File: cmd/tailscale/cli/switch.go [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+	"time"
+
+	"github.com/peterbourgon/ff/v3/ffcli"
+	"tailscale.com/ipn"
+)
+
+var switchCmd = &ffcli.Command{
+	Name:      "switch",
+	ShortHelp: "Switches to a different Tailscale profile",
+	// ShortUsage: "switch <name>\r\n  switch --list",
```
Comment (bradfitz) [inline cmd/tailscale/cli/switch.go:21]: ?

File: cmd/tailscale/cli/switch.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	}
+	var profID ipn.ProfileID
+	for _, p := range all {
+		if p.Name == args[0] {
+			profID = p.ID
+			break
+		}
+	}
+	if profID == "" {
+		printf("No profile named %q\n", args[0])
```
Comment (bradfitz) [inline cmd/tailscale/cli/switch.go:77]: to stdout and not stderr for errors? (here and elsewhere)

File: cmd/tailscale/cli/switch.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+		st, err := localClient.StatusWithoutPeers(ctx)
+		if err != nil {
+			outln("Error getting status:", err)
+			os.Exit(1)
+		}
+		switch st.BackendState {
+		case "NoState", "Starting":
+			// TODO(maisem): maybe add a way to subscribe to state changes to
+			// LocalClient.
+			time.Sleep(100 * time.Millisecond)
```
Comment (bradfitz) [inline cmd/tailscale/cli/switch.go:105]: ouch. but yeah, I want to add the non-HTTP IPN protocol's streaming to a Watch method on LocalClient and then remove the non-HTTP protocol entirely.

**Review by bradfitz — APPROVED** [review]

File: cmd/tailscale/cli/switch.go (hunk 10 above)
Comment (maisem) [reply cmd/tailscale/cli/switch.go]: debug

File: cmd/tailscale/cli/cli.go (hunk 8 above)
Comment (maisem) [reply cmd/tailscale/cli/cli.go]: Done

File: client/tailscale/localclient.go (hunk 7 above)
Comment (maisem) [reply client/tailscale/localclient.go]: Done

File: client/tailscale/localclient.go (hunk 6 above)
Comment (maisem) [reply client/tailscale/localclient.go]: Done.

**Pull request merged**